synapse/handlers/oidc_handler.py

```diff
@@ -906,20 +908,25 @@ async def _map_userinfo_to_user(
         localpart = map_username_to_mxid_localpart(attributes["localpart"])
 
         user_id = UserID(localpart, self._hostname)
-        if await self._datastore.get_users_by_id_case_insensitive(user_id.to_string()):
-            # This mxid is taken
-            raise MappingException(
-                "mxid '{}' is already taken".format(user_id.to_string())
-            )
-
-        # It's the first time this user is logging in and the mapped mxid was
-        # not taken, register the user
-        registered_user_id = await self._registration_handler.register_user(
-            localpart=localpart,
-            default_display_name=attributes["display_name"],
-            user_agent_ips=(user_agent, ip_address),
+        matches = await self._datastore.get_users_by_id_case_insensitive(
+            user_id.to_string()
         )
-
+        if matches:
+            if self._allow_existing_users:
+                registered_user_id = next(iter(matches))
```

Comment:
There's an open conversation about this on the old PR: https://github.com/matrix-org/synapse/pull/8180#discussion_r490488714

The gist is:

* `get_users_by_id_case_insensitive` returns a dictionary of username -> password hash.
* Can this ever have more than one result in it since it seems that usernames are case-insensitive?
* [`AuthHandler._find_user_id_and_pwd_hash`](https://github.com/matrix-org/synapse/blob/efb6b6629c78409251f61857f2bfe6c2f8f8fb8d/synapse/handlers/auth.py#L753-L758) has logic to handle multiple matches.

I wonder if this should use similar logic to `AuthHandler._find_user_id_and_pwd_hash`? We probably don't want to use that since it logs a warning if no matches are found, but we actually expect that in some situations!

Reply:
It seems [the spec](https://matrix.org/docs/spec/appendices#user-identifiers) does say they are case insensitive. I don't know if this is just a guard against the database being "wrong"?

Reply:
I find `SamlHandler` also simply uses the first match.
https://github.com/matrix-org/synapse/blob/aec294ee0d0f2fa4ccef57085d670b8939de3669/synapse/handlers/saml_handler.py#L307-L314

In fact, these two ways are different. Using the first match means case insensitive, you can log in `user` with `USER`. Checking if it in this dictionary means case sensitive but usernames which differ only in case are not allowed. Both of them are a little different from the spec, which just disallows upper-case characters.

I suggest that at least we should keep things the same.

Reply:
We had a quick chat about this in [#synapse-dev:matrix.org](https://matrix.to/#/!XaqDhxuTIlvldquJaV:matrix.org/$-J9eYk-of24OMLHrFX74zLtLaRULd9N8aCTkEOFIh4Y?via=matrix.org&via=vector.modular.im&via=privacytools.io) and the conclusion was:

* Usernames used to be case-insensitive, but are now case-sensitive. There's some code to magically handle this during login.
* The general logic (which seems to be what `AuthHandler._find_user_id_and_pwd_hash` uses):
    1. If there's an exact match, use it.
    2. If there's a single case-insensitive match, use it.
    3. Otherwise, there's nothing to do automatically.

It seems that both the SAML code and OIDC code should use this to be consistent.

Reply:
Should I convert the single case-insensitive match, which is a `string`, to `UserID` object?
Maybe I should use `registered_user_id = user_id.from_string(UserId, next(iter(users)))` ? 
(I'm not sure if it's the right way to use `DomainSpecificString.from_string`)